In [1]:
import polars as pl
import pandas as pd
from IPython.display import display

In [2]:
from Predictors.Pb.features import validate_features_vs_target

In [3]:
from data.Constants.ColumnNames import WsfrlColumnOverall
from data.Constants.ColumnNames import WsfrlColumnPriorStats
from data.Constants.ColumnNames import WsfrlColumnForAges


In [4]:
import Predictors.DataPrep as dp
df_ages = dp.get_data()

display(df_ages)

shape: (72_219, 74)
┌──────────┬──────────┬────────────┬────────┬───┬────────────┬────────────┬────────────┬───────────┐
│ Position ┆ Time     ┆ time_in_se ┆ Bib    ┆ … ┆ p_of_10_ro ┆ p_of_10_xc ┆ IsAgeCateg ┆ is_big_cl │
│ ---      ┆ ---      ┆ conds      ┆ Number ┆   ┆ ad_age_cat ┆ _age_categ ┆ oryMissing ┆ ub        │
│ i64      ┆ str      ┆ ---        ┆ ---    ┆   ┆ egory_as_o ┆ ory_as_of_ ┆ ---        ┆ ---       │
│          ┆          ┆ f64        ┆ i64    ┆   ┆ …          ┆ …          ┆ bool       ┆ bool      │
│          ┆          ┆            ┆        ┆   ┆ ---        ┆ ---        ┆            ┆           │
│          ┆          ┆            ┆        ┆   ┆ str        ┆ str        ┆            ┆           │
╞══════════╪══════════╪════════════╪════════╪═══╪════════════╪════════════╪════════════╪═══════════╡
│ 1        ┆ 00:27:42 ┆ 1662.0     ┆ 325    ┆ … ┆ Age        ┆ Age        ┆ true       ┆ false     │
│ 218      ┆ 00:48:20 ┆ 2900.0     ┆ 328    ┆ … ┆ Age        ┆ Age        ┆ true       ┆ true      │
│ 217      ┆ 00:48:17 ┆ 2897.0     ┆ 182    ┆ … ┆ V65        ┆ V65        ┆ false      ┆ true      │
│ 216      ┆ 00:48:12 ┆ 2892.0     ┆ 234    ┆ … ┆ V60        ┆ V60        ┆ false      ┆ true      │
│ 215      ┆ 00:48:08 ┆ 2888.0     ┆ 27     ┆ … ┆ V50        ┆ V50        ┆ false      ┆ true      │
│ …        ┆ …        ┆ …          ┆ …      ┆ … ┆ …          ┆ …          ┆ …          ┆ …         │
│ 105      ┆ 00:36:17 ┆ 2177.0     ┆ 228    ┆ … ┆ V55        ┆ V55        ┆ false      ┆ true      │
│ 104      ┆ 00:36:14 ┆ 2174.0     ┆ 160    ┆ … ┆ V60        ┆ V60        ┆ false      ┆ true      │
│ 103      ┆ 00:36:05 ┆ 2165.0     ┆ 224    ┆ … ┆ Age        ┆ Age        ┆ true       ┆ true      │
│ 121      ┆ 00:37:29 ┆ 2249.0     ┆ 269    ┆ … ┆ V60        ┆ V60        ┆ false      ┆ true      │
│ 325      ┆ 01:09:43 ┆ 4183.0     ┆ 155    ┆ … ┆ V50        ┆ V50        ┆ false      ┆ true      │
└──────────┴──────────┴────────────┴────────┴───┴────────────┴────────────┴────────────┴───────────┘

In [5]:
# Remove all rows where the year is 2014
df_ages = df_ages.filter(
     (pl.col(WsfrlColumnForAges.AgeCategoryRoad) != "TBC")
    # & (pl.col('year') != 2014)
    & pl.col(WsfrlColumnForAges.AgeAvgRange).is_not_null()
    & pl.col(WsfrlColumnForAges.AgeMinRange).is_not_null()
    & pl.col(WsfrlColumnForAges.AgeMaxRange).is_not_null()
)


In [6]:
 # Pivot the DataFrame
pivot_df_year_by_age_cat = df_ages.pivot(
    values=[WsfrlColumnForAges.AgeCategoryRoad],
    index="year",
    on=WsfrlColumnForAges.AgeCategoryRoad,
    aggregate_function="len"
).sort("year")

display(pivot_df_year_by_age_cat)


shape: (13, 19)
┌──────┬─────┬─────┬─────┬───┬──────┬──────┬──────┬──────┐
│ year ┆ V50 ┆ V45 ┆ V35 ┆ … ┆ U13  ┆ U23  ┆ U15  ┆ V80  │
│ ---  ┆ --- ┆ --- ┆ --- ┆   ┆ ---  ┆ ---  ┆ ---  ┆ ---  │
│ i64  ┆ u32 ┆ u32 ┆ u32 ┆   ┆ u32  ┆ u32  ┆ u32  ┆ u32  │
╞══════╪═════╪═════╪═════╪═══╪══════╪══════╪══════╪══════╡
│ 2011 ┆ 466 ┆ 410 ┆ 255 ┆ … ┆ null ┆ null ┆ null ┆ null │
│ 2012 ┆ 379 ┆ 405 ┆ 226 ┆ … ┆ null ┆ null ┆ null ┆ null │
│ 2013 ┆ 499 ┆ 528 ┆ 329 ┆ … ┆ null ┆ null ┆ null ┆ null │
│ 2014 ┆ 720 ┆ 700 ┆ 565 ┆ … ┆ 7    ┆ null ┆ null ┆ null │
│ 2015 ┆ 786 ┆ 797 ┆ 624 ┆ … ┆ 3    ┆ null ┆ null ┆ null │
│ …    ┆ …   ┆ …   ┆ …   ┆ … ┆ …    ┆ …    ┆ …    ┆ …    │
│ 2019 ┆ 851 ┆ 781 ┆ 342 ┆ … ┆ null ┆ null ┆ 3    ┆ 21   │
│ 2020 ┆ 62  ┆ 58  ┆ 34  ┆ … ┆ 1    ┆ null ┆ null ┆ null │
│ 2022 ┆ 583 ┆ 401 ┆ 215 ┆ … ┆ 7    ┆ null ┆ 2    ┆ 12   │
│ 2023 ┆ 524 ┆ 463 ┆ 203 ┆ … ┆ null ┆ 26   ┆ null ┆ 14   │
│ 2024 ┆ 663 ┆ 470 ┆ 268 ┆ … ┆ 8    ┆ 26   ┆ 15   ┆ 9    │
└──────┴─────┴─────┴─────┴───┴──────┴──────┴──────┴──────┘

In [7]:
 # Pivot the DataFrame
# Assuming df_ages is your Polars DataFrame
pivot_df_age_cat = df_ages.pivot(
    values=[WsfrlColumnForAges.AgeCategoryRoad],  # Use "*" to count all occurrences
    index=WsfrlColumnForAges.AgeCategoryRoad,
    on=WsfrlColumnOverall.Gender,
    aggregate_function="len"
).to_pandas()

print(type(df_ages))
# # Sum the counts across gender columns to create a row total
# pivot_df = pivot_df.with_column(
#     pl.sum_horizontal(pivot_df.select(pl.col(WsfrlColumnOverall.Gender).unique())).alias("row_total")
# )
pivot_df_age_cat["All"] = pivot_df_age_cat["M"] + pivot_df_age_cat["F"] 

pivot_df_age_cat = pivot_df_age_cat.sort_values(by="p_of_10_road_age_category", ascending=False, inplace=True)



<class 'polars.dataframe.frame.DataFrame'>


In [8]:
from data.Constants.ColumnNames import WsfrlColumnForAgesGenerated

age_category_mapping = {
    "SEN": "Senior",
    "TBC": "Junior",
    "U11": "Junior",
    "U13": "Junior",
    "U15": "Junior",
    "U17": "Junior",
    "U20": "Junior",
    "U23": "Junior",
    "V35": "Senior",
    "V40": "Veteran V40",
    "V45": "Veteran V45",
    "V50": "Veteran V50",
    "V55": "Veteran V55",
    "V60": "Veteran V60",
    "V65": "Veteran V65",
    "V70": "Veteran V70+",
    "V75": "Veteran V70+",
    "V80": "Veteran V70+",
    "V85": "Veteran V70+",
    "V95": "Veteran V70+"
}

age_category_mapping_plus = {
    "SEN": "Senior",
    "TBC": "Junior",
    "U11": "Junior",
    "U13": "Junior",
    "U15": "Junior",
    "U17": "Junior",
    "U20": "Junior",
    "U23": "Junior",
    "V35": "Senior",
    "V40": "Veteran V40+",
    "V45": "Veteran V45+",
    "V50": "Veteran V50+",
    "V55": "Veteran V55+",
    "V60": "Veteran V60+",
    "V65": "Veteran V65+",
    "V70": "Veteran V65+",
    "V75": "Veteran V65+",
    "V80": "Veteran V65+",
    "V85": "Veteran V65+",
    "V95": "Veteran V65+"
}


df_ages = df_ages.with_columns((pl.col("p_of_10_road_age_category").map_elements(lambda x: age_category_mapping.get(x), return_dtype=pl.String).alias(str(WsfrlColumnForAgesGenerated.BlendedAgeCategory))))

df_ages = df_ages.with_columns((pl.col("p_of_10_road_age_category").map_elements(lambda x: age_category_mapping_plus.get(x), return_dtype=pl.String).alias(str(WsfrlColumnForAgesGenerated.BlendedAgeCategoryPlus))))



In [10]:
# Pivot the DataFrame
# Assuming df_ages is your Polars DataFrame
pivot_df_blended_age_cat = df_ages.pivot(
    values=[WsfrlColumnForAgesGenerated.BlendedAgeCategory],  # Use "*" to count all occurrences
    index=[WsfrlColumnForAgesGenerated.BlendedAgeCategoryPlus,WsfrlColumnForAgesGenerated.BlendedAgeCategory],
    on=WsfrlColumnOverall.Gender,
    aggregate_function="len"
).to_pandas()

pivot_df_blended_age_cat = pivot_df_blended_age_cat.sort_values(
    by=[WsfrlColumnForAgesGenerated.BlendedAgeCategoryPlus, WsfrlColumnForAgesGenerated.BlendedAgeCategory]
)

print(type(df_ages))
# # Sum the counts across gender columns to create a row total
# pivot_df = pivot_df.with_column(
#     pl.sum_horizontal(pivot_df.select(pl.col(WsfrlColumnOverall.Gender).unique())).alias("row_total")
# )
pivot_df_blended_age_cat["All"] = pivot_df_blended_age_cat["M"] + pivot_df_blended_age_cat["F"] 

pivot_df_blended_age_cat

<class 'polars.dataframe.frame.DataFrame'>


blended_age_category_plus blended_age_category     M     F   All
7                    Junior               Junior   834   621  1455
2                    Senior               Senior  5267  3589  8856
3              Veteran V40+          Veteran V40  3129  2455  5584
1              Veteran V45+          Veteran V45  4004  3103  7107
0              Veteran V50+          Veteran V50  4486  3352  7838
4              Veteran V55+          Veteran V55  4375  2759  7134
6              Veteran V60+          Veteran V60  3753  1728  5481
5              Veteran V65+          Veteran V65  2310   686  2996
8              Veteran V65+         Veteran V70+  1859   455  2314

## Age Category

Dilemma is that the age category is  unbalanced
- Junior are quite time
- Seniors are large and cannot be broken down further


| Blended Cat     | Sum of All | % of total |
|-----------------|------------:|-----------:|
| Junior          |       1455 |          3% |
| Senior          |       8856 |         18% |
| Veteran V40+    |       5584 |         11% |
| Veteran V45+    |       7107 |         15% |
| Veteran V50+    |       7838 |         16% |
| Veteran V55+    |       7134 |         15% |
| Veteran V60+    |       5481 |         11% |
| Veteran V65+    |       5310 |         11% |
| **Grand Total** | **48765**  |     **100%** |	


In [ ]:
# Remove all junior as they are a tiny sub population 
# I end up keeping them 
# df_ages = df_ages.filter(
#      (pl.col(WsfrlColumnForAgesGenerated.BlendedAgeCategory) != "Junior")
# )

In [11]:
target_columns = [
    str(WsfrlColumnForAges.AgeAvgRange), 
    str(WsfrlColumnForAges.AgeMinRange), 
    str(WsfrlColumnForAges.AgeMaxRange),
    str(WsfrlColumnForAges.AgeCategoryRoad),
    str(WsfrlColumnForAgesGenerated.BlendedAgeCategory),
    str(WsfrlColumnForAgesGenerated.BlendedAgeCategoryPlus)
    ]



WsfrlColumnCalculated.PriorVsRun = "UP" "Down"


In [12]:
df_ages = df_ages.to_dummies(
    [WsfrlColumnOverall.Gender, WsfrlColumnOverall.Club], drop_first=True 
)
display(df_ages)

df_ages.to_pandas()

shape: (48_765, 187)
┌──────────┬──────────┬────────────┬────────┬───┬────────────┬────────────┬────────────┬───────────┐
│ Position ┆ Time     ┆ time_in_se ┆ Bib    ┆ … ┆ IsAgeCateg ┆ is_big_clu ┆ blended_ag ┆ blended_a │
│ ---      ┆ ---      ┆ conds      ┆ Number ┆   ┆ oryMissing ┆ b          ┆ e_category ┆ ge_catego │
│ i64      ┆ str      ┆ ---        ┆ ---    ┆   ┆ ---        ┆ ---        ┆ ---        ┆ ry_plus   │
│          ┆          ┆ f64        ┆ i64    ┆   ┆ bool       ┆ bool       ┆ str        ┆ ---       │
│          ┆          ┆            ┆        ┆   ┆            ┆            ┆            ┆ str       │
╞══════════╪══════════╪════════════╪════════╪═══╪════════════╪════════════╪════════════╪═══════════╡
│ 217      ┆ 00:48:17 ┆ 2897.0     ┆ 182    ┆ … ┆ false      ┆ true       ┆ Veteran    ┆ Veteran   │
│          ┆          ┆            ┆        ┆   ┆            ┆            ┆ V50        ┆ V50+      │
│ 216      ┆ 00:48:12 ┆ 2892.0     ┆ 234    ┆ … ┆ false      ┆ true       ┆ Veteran    ┆ Veteran   │
│          ┆          ┆            ┆        ┆   ┆            ┆            ┆ V45        ┆ V45+      │
│ 215      ┆ 00:48:08 ┆ 2888.0     ┆ 27     ┆ … ┆ false      ┆ true       ┆ Senior     ┆ Senior    │
│ 214      ┆ 00:48:06 ┆ 2886.0     ┆ 241    ┆ … ┆ false      ┆ true       ┆ Veteran    ┆ Veteran   │
│          ┆          ┆            ┆        ┆   ┆            ┆            ┆ V50        ┆ V50+      │
│ 213      ┆ 00:48:05 ┆ 2885.0     ┆ 292    ┆ … ┆ false      ┆ true       ┆ Veteran    ┆ Veteran   │
│          ┆          ┆            ┆        ┆   ┆            ┆            ┆ V40        ┆ V40+      │
│ …        ┆ …        ┆ …          ┆ …      ┆ … ┆ …          ┆ …          ┆ …          ┆ …         │
│ 106      ┆ 00:36:22 ┆ 2182.0     ┆ 320    ┆ … ┆ false      ┆ true       ┆ Veteran    ┆ Veteran   │
│          ┆          ┆            ┆        ┆   ┆            ┆            ┆ V50        ┆ V50+      │
│ 105      ┆ 00:36:17 ┆ 2177.0     ┆ 228    ┆ … ┆ false      ┆ true       ┆ Veteran    ┆ Veteran   │
│          ┆          ┆            ┆        ┆   ┆            ┆            ┆ V55        ┆ V55+      │
│ 104      ┆ 00:36:14 ┆ 2174.0     ┆ 160    ┆ … ┆ false      ┆ true       ┆ Veteran    ┆ Veteran   │
│          ┆          ┆            ┆        ┆   ┆            ┆            ┆ V60        ┆ V60+      │
│ 121      ┆ 00:37:29 ┆ 2249.0     ┆ 269    ┆ … ┆ false      ┆ true       ┆ Veteran    ┆ Veteran   │
│          ┆          ┆            ┆        ┆   ┆            ┆            ┆ V60        ┆ V60+      │
│ 325      ┆ 01:09:43 ┆ 4183.0     ┆ 155    ┆ … ┆ false      ┆ true       ┆ Veteran    ┆ Veteran   │
│          ┆          ┆            ┆        ┆   ┆            ┆            ┆ V50        ┆ V50+      │
└──────────┴──────────┴────────────┴────────┴───┴────────────┴────────────┴────────────┴───────────┘

Position      Time  time_in_seconds  Bib Number             Name  \
0           217  00:48:17           2897.0         182       Paul Crowe   
1           216  00:48:12           2892.0         234  Nigel Mcnestrie   
2           215  00:48:08           2888.0          27     Linda Cooper   
3           214  00:48:06           2886.0         241    Gary Tompsett   
4           213  00:48:05           2885.0         292  Michelle Hughes   
...         ...       ...              ...         ...              ...   
48760       106  00:36:22           2182.0         320    Darren Winter   
48761       105  00:36:17           2177.0         228      Steve Jones   
48762       104  00:36:14           2174.0         160     Mark Kennard   
48763       121  00:37:29           2249.0         269      Evert Baker   
48764       325  01:09:43           4183.0         155  Cheryl Costello   

       Gender_F  Gender Position  Club_Almost Athletes  Club_Arena 80 AC  \
0             0            170.0                     0                 0   
1             0            169.0                     0                 0   
2             1             47.0                     0                 0   
3             0            168.0                     0                 0   
4             1             46.0                     0                 0   
...         ...              ...                   ...               ...   
48760         0             90.0                     0                 0   
48761         0             89.0                     0                 0   
48762         0             88.0                     0                 0   
48763         0             99.0                     0                 0   
48764         1            138.0                     0                 0   

       Club_Arunners  ...  min_dob  max_dob  avg_dob  \
0                  0  ...   1959.0   1955.0   1957.0   
1                  0  ...   1964.0   1960.0   1962.0   
2                  1  ...   1974.0   1970.0   1972.0   
3                  0  ...   1959.0   1955.0   1957.0   
4                  0  ...   1969.0   1965.0   1967.0   
...              ...  ...      ...      ...      ...   
48760              0  ...   1974.0   1970.0   1972.0   
48761              0  ...   1969.0   1965.0   1967.0   
48762              0  ...   1964.0   1960.0   1962.0   
48763              0  ...   1964.0   1960.0   1962.0   
48764              0  ...   1974.0   1970.0   1972.0   

       p_of_10_track_age_category_as_of_2024  \
0                                        V65   
1                                        V60   
2                                        V50   
3                                        V65   
4                                        V55   
...                                      ...   
48760                                    V50   
48761                                    V55   
48762                                    V60   
48763                                    V60   
48764                                    V50   

       p_of_10_road_age_category_as_of_2024  \
0                                       V65   
1                                       V60   
2                                       V50   
3                                       V65   
4                                       V55   
...                                     ...   
48760                                   V50   
48761                                   V55   
48762                                   V60   
48763                                   V60   
48764                                   V50   

       p_of_10_xc_age_category_as_of_2024  IsAgeCategoryMissing  is_big_club  \
0                                     V65                 False         True   
1                                     V60                 False         True   
2                                     V50                 False         True   
3                                     V65                 

## Feature Selection




In [19]:
from Predictors.Pb.features import get_boolean_features

feature_columns_club = get_boolean_features(df_ages, 'Club_')


In [20]:

feature_columns_to_default_to_zero = [
    str(WsfrlColumnPriorStats.Tenure),
    str(WsfrlColumnPriorStats.TotalPriorRunCount),
    str(WsfrlColumnPriorStats.PositionYearsSinceBest),
    str(WsfrlColumnPriorStats.PositionYearSinceWorst),
]


feature_columns_mandatory = [
    str(WsfrlColumnOverall.Position),
    f"{WsfrlColumnOverall.Gender}_F",
    str(WsfrlColumnOverall.TimeInSeconds),
    str(WsfrlColumnOverall.IsClubMember),
    str(WsfrlColumnOverall.IsBigClub),
]



feature_columns = (feature_columns_to_default_to_zero 
                   + feature_columns_mandatory 
                   # + feature_columns_club                 
                   )




AttributeError: type object 'WsfrlColumnOverall' has no attribute 'IsClubMember'

In [ ]:

df_features = df_ages[feature_columns]
df_target = df_ages[target_columns]


validate_features_vs_target(df_features, df_target)

# Display the reduced DataFrame
# display(df_features.head())
# display(df_target.head())



## Split the Data

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

SEED = 123

(
    df_feature_train,
    df_feature_test,
    df_target_train,
    df_target_test,
) = train_test_split(
    df_features,
    df_target,
    test_size=0.30,
    random_state=SEED,
)

total_rows = len(df_features)
train_rows = len(df_feature_train)
test_rows = len(df_feature_test)

test_percentage_of_total = train_rows / total_rows

print(f"Total Rows: {total_rows} Training Rows: {train_rows} Test Rows: {test_rows} Test Percentage of Total: {test_percentage_of_total:.2f}")

# train_test_split return numpy arrays, convert them back to Pandas DataFrames
# Because that is what the pipeline expects
df_feature_train.to_pandas(),
df_feature_test.to_pandas(),
df_target_train.to_pandas(),
df_target_test.to_pandas(),

validate_features_vs_target(df_feature_train,df_target_train)
validate_features_vs_target(df_feature_test,df_target_test)


In [ ]:
from Predictors.Pb.features import general_feature_validation

general_feature_validation(df_feature_train, feature_columns)

In [ ]:
from enum import Enum
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier

class RegressionMethod(Enum):
    LOGISTIC = "LogisticRegression"
    LINEAR = "LinearRegression"
    DECISIONTREE = "DecisionTree"



def build_pipeline(target_column: str, method: RegressionMethod ) -> Pipeline:
    # Create imputers for different features/columns
    default_to_zero_imputer = SimpleImputer(strategy="constant", fill_value=0) #-np.inf is too big for a float
    binary_value_imputer = SimpleImputer(strategy="constant", fill_value=np.uint8(0))
    mandatory_average_imputer = SimpleImputer(strategy="constant", fill_value=0)
    
    
    # Combine the imputers using ColumnTransformer
    preprocessor = ColumnTransformer(
        transformers=[
            ("default_zero_features", default_to_zero_imputer, feature_columns_to_default_to_zero),
            # ("club_features", binary_value_imputer, feature_columns_club),
            ("mandatory_features", mandatory_average_imputer, feature_columns_mandatory),
        ],
        # remainder="drop",
    )
    
    PIPELINE_STEP_PREPROCESSOR = "preprocessor"
    PIPELINE_STEP_SCALER = "scaler"
    PIPELINE_STEP_CLASSIFIER = "classifier"
    
    # Create a pipeline with the preprocessor, scaler, and the appropriate regression model
    steps = [
        (PIPELINE_STEP_PREPROCESSOR, preprocessor),
        (PIPELINE_STEP_SCALER, StandardScaler()),
    ]
    
    if method == RegressionMethod.LOGISTIC:
        steps.append((PIPELINE_STEP_CLASSIFIER, LogisticRegression(penalty=None, max_iter=1000)))
    elif method == RegressionMethod.LINEAR:
        steps.append((PIPELINE_STEP_CLASSIFIER, LinearRegression()))
    elif method == RegressionMethod.DECISIONTREE:
        steps.append((PIPELINE_STEP_CLASSIFIER, LinearRegression()))
    
    pipeline = Pipeline(steps=steps)

    # Fit the pipeline on the training data
    pipeline.fit(df_feature_train, df_target_train[target_column])
    return pipeline



In [ ]:
pipeline_logistic_regression = (
    build_pipeline(WsfrlColumnForAges.AgeCategoryRoad, RegressionMethod.LOGISTIC))

pipeline_logistic_regression_blended_cat = (
    build_pipeline(WsfrlColumnForAgesGenerated.BlendedAgeCategory, RegressionMethod.LOGISTIC))

pipeline_logistic_regression_blended_cat_plus = (
    build_pipeline(WsfrlColumnForAgesGenerated.BlendedAgeCategoryPlus, RegressionMethod.LOGISTIC))

pipeline_linear_regression = build_pipeline(
    WsfrlColumnForAges.AgeAvgRange, RegressionMethod.LINEAR)

In [ ]:

def run_pipeline(pipeline: Pipeline, df_feature: pd.DataFrame):
    # Transform the test data using the preprocessor part of the pipeline
    preprocessed_df = pipeline.named_steps['preprocessor'].transform(df_feature)

    # Scale the preprocessed test data
    scaled_df = pipeline.named_steps['scaler'].transform(preprocessed_df)
    
    return scaled_df
    
def run_logistic_regression_pipeline_in_stages(pipeline: Pipeline,  
                                               df_feature: pd.DataFrame, 
                                               df_target: pd.DataFrame, 
                                               target_column: str) -> tuple:
    
    # Scale the preprocessed test data
    scaled_df = run_pipeline(pipeline, df_feature)

    classifier = LogisticRegression(penalty=None, max_iter=1000)

    classifier.fit(
        scaled_df,
        df_target[target_column],
    )

    return scaled_df, classifier



## Age Category
Use logistic regression to predict the **age category** which is **very UNbalanced**

In [ ]:

(scaled_train_df_age_cat,
logistic_regression_classifier) = run_logistic_regression_pipeline_in_stages(
    pipeline_logistic_regression,
    df_feature_train,
    df_target_train,
    WsfrlColumnForAges.AgeCategoryRoad )





## blended category
Use logistic regression to predict the **blended category** which is **very balanced**

In [ ]:
(scaled_train_df_age_cat_blended,
logistic_regression_classifier_for_blended) = run_logistic_regression_pipeline_in_stages(
    pipeline_logistic_regression_blended_cat,
    df_feature_train,
    df_target_train,
    WsfrlColumnForAgesGenerated.BlendedAgeCategory )


In [ ]:
(scaled_train_df_age_cat_blended_plus,
logistic_regression_classifier_for_blended_plus) = run_logistic_regression_pipeline_in_stages(
    pipeline_logistic_regression_blended_cat_plus,
    df_feature_train,
    df_target_train,
    WsfrlColumnForAgesGenerated.BlendedAgeCategoryPlus)

In [ ]:
def run_linear_regression_pipeline_in_stages(pipeline: Pipeline,  
                                               df_feature: pd.DataFrame, 
                                               df_target: pd.DataFrame, 
                                               target_column: str) -> tuple:
    
    # Scale the preprocessed test data
    scaled_df = run_pipeline(pipeline, df_feature)

   # Create a LinearRegression model
    regressor = LinearRegression()
    
    # Fit the regressor on the scaled training data
    regressor.fit(
        scaled_df,
        df_target[target_column],
    )

    return scaled_df, regressor

In [ ]:
(scaled_train_df_avg_age,
linear_regression) = run_linear_regression_pipeline_in_stages(
    pipeline_linear_regression,
    df_feature_train,
    df_target_train,
    WsfrlColumnForAges.AgeAvgRange )

In [ ]:
import altair as alt

def display_feature_coefficients(classifier, feature_columns):
    # Convert the set to a list to maintain order
    all_feature_columns_list = list(feature_columns)
    
    
    # Create a DataFrame with the features and their coefficients
    feature_coefficients_df = pl.DataFrame(
        {
            "feature": all_feature_columns_list,
            "coefficient": classifier.coef_[0][:len(feature_columns)],
        }
    )
    
    # Sort the DataFrame by coefficient in descending order
    sorted_feature_coefficients_df = feature_coefficients_df.sort("coefficient", descending=True)
    
    # Plot the bar chart
    alt.Chart(sorted_feature_coefficients_df.to_pandas()).mark_bar().encode(
        x="coefficient",
        y=alt.Y("feature", sort="-x")
    ).properties(
        width=500,
        height=300,
    ).display()

In [ ]:
display_feature_coefficients(logistic_regression_classifier, feature_columns)

In [ ]:
display_feature_coefficients(logistic_regression_classifier_for_blended, feature_columns)

In [ ]:
display_feature_coefficients(logistic_regression_classifier_for_blended_plus, feature_columns)

In [ ]:
# Make predictions on the test data
predictions = pipeline_linear_regression.named_steps['classifier'].predict(scaled_train_df_avg_age)

## Unbalnced (age Cat)


In [ ]:
(scaled_test_df_age_cat,
junk__as_on_test_logistic_regression_classifier) = run_logistic_regression_pipeline_in_stages(
    pipeline_logistic_regression,
    df_feature_test,
    df_target_test,
    WsfrlColumnForAges.AgeCategoryRoad )

In [ ]:
(scaled_test_df_age_cat_blended,
still_junk_as_on_test_logistic_regression_classifier) = run_logistic_regression_pipeline_in_stages(
    pipeline_logistic_regression_blended_cat,
    df_feature_test,
    df_target_test,
    WsfrlColumnForAgesGenerated.BlendedAgeCategory)

In [ ]:
(scaled_test_df_age_cat_blended_plus,
still_junk_as_on_test_logistic_regression_classifier_plus) = run_logistic_regression_pipeline_in_stages(
    pipeline_logistic_regression_blended_cat_plus,
    df_feature_test,
    df_target_test,
    WsfrlColumnForAgesGenerated.BlendedAgeCategoryPlus)

In [ ]:
## Balnced (blended)

In [ ]:
from sklearn.metrics import classification_report

# Must be scaled like training data
validate_features_vs_target(scaled_test_df_age_cat, df_feature_test)

# Cat Age (unbalanced)
print(
    classification_report(
        df_target_test.get_column(WsfrlColumnForAges.AgeCategoryRoad),
        logistic_regression_classifier.predict(scaled_test_df_age_cat),
        zero_division=0
    )
)

# Blended Cat Age (balanced)
print(
    classification_report(
        df_target_test.get_column(WsfrlColumnForAgesGenerated.BlendedAgeCategory),
        logistic_regression_classifier_for_blended.predict(scaled_test_df_age_cat_blended),
        zero_division=0
    )
)

# Blended Cat Age PLUS ( high level and balanced)
print(
    classification_report(
        df_target_test.get_column(WsfrlColumnForAgesGenerated.BlendedAgeCategoryPlus),
        logistic_regression_classifier_for_blended_plus.predict(scaled_test_df_age_cat_blended_plus),
        zero_division=0
    )
)

## Really poor to predict

did not do very well

> Overall Accuracy:
    The overall accuracy is 0.20 (20%), meaning the model's predictions are correct 20% of the time.
    This is low, especially if compared to a baseline model (e.g., always predicting the most common class). 

> Classes like SEN, V45, V50, V55, and V60 dominate the dataset, while smaller classes like U11, U13, and U15 have minimal representation. Logistic regression may overly focus on larger classes, neglecting smaller ones.

In [ ]:
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier, plot_tree

#scaled_test_df_age_cat_blended_plus

def run_decision_tree_pipeline_in_stages(pipeline: Pipeline,  
                                               df_feature: pd.DataFrame, 
                                               df_target: pd.DataFrame, 
                                               target_column: str) -> tuple:
    
    # Scale the preprocessed test data
    scaled_df = run_pipeline(pipeline, df_feature)

    classifier = DecisionTreeClassifier(max_depth=3, random_state=SEED)
     
    classifier.fit(
        scaled_df,
        df_target[target_column],
    )
    
    return scaled_df, classifier
    


In [ ]:
(tree_scaled_train_df_age_cat_blended_plus,
 decision_tree_classifier_for_blended_plus
 ) = run_decision_tree_pipeline_in_stages(
    pipeline_logistic_regression_blended_cat_plus,
    df_feature_train,
    df_target_train,
    WsfrlColumnForAgesGenerated.BlendedAgeCategoryPlus)

In [ ]:
class_names = sorted(df_ages[WsfrlColumnForAgesGenerated.BlendedAgeCategoryPlus].unique())

plt.figure(figsize=(10, 10), dpi=300)
plot_tree(
    decision_tree_classifier_for_blended_plus, feature_names=df_feature_train.columns, class_names=class_names, filled=True
)
plt.show()

In [ ]:
(scaled_test_df_avg_age,
junk_as_on_test_linear_regression) = run_linear_regression_pipeline_in_stages(
    pipeline_linear_regression,
    df_feature_test,
    df_target_test,
    WsfrlColumnForAges.AgeAvgRange )

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Make predictions on the test data
predictions = linear_regression.predict(scaled_test_df_avg_age)

# Calculate evaluation metrics
mae = mean_absolute_error(df_target_test[WsfrlColumnForAges.AgeAvgRange], predictions)
mse = mean_squared_error(df_target_test[WsfrlColumnForAges.AgeAvgRange], predictions)
r2 = r2_score(df_target_test[WsfrlColumnForAges.AgeAvgRange], predictions)

print(f"Mean Absolute Error (MAE): {mae}")
print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R²): {r2}")